lib/web_of_science/links_client.rb

```diff
@@ -0,0 +1,85 @@
+module WebOfScience
+  # Links AMR (Article Match Retrieval) Service
+  # @see http://ipscience-help.thomsonreuters.com/LAMRService/WebServicesOverviewGroup/overview.html Service documentation
+  class LinksClient
+
+    LINKS_HOST = 'https://ws.isiknowledge.com'.freeze
+    LINKS_PATH = '/cps/xrpc'.freeze
+
+    attr_reader :username, :password, :host
+
+    # @param [String] username
+    # @param [String] password
+    # @param [String] host
+    def initialize(username: nil, password: nil, host: LINKS_HOST)
+      @host = host
+      @username = username
+      @password = password
+      (@username, @password) = Base64.decode64(Settings.WOS.AUTH_CODE).split(':', 2) unless username
+    end
+
+    # Retrieve identifier 'fields' for the record 'ids'
+    # @param [Array<String>] ids
+    # @param [Array<String>] fields (defaults to ['ut', 'doi', 'pmid'])
+    # @return [Hash<String => Hash>]
+    def links(ids, fields = %w(ut doi pmid))
+      ids.uniq!
+      raise ArgumentError, '1-50 ids required' if ids.empty? || ids.count > 50
+      raise ArgumentError, 'fields cannot be empty' if fields.empty?
+      response = connection.post do |req|
+        req.path = LINKS_PATH
+        req.body = request_body(ids, fields)
+      end
+      response_parse(response)
+    end
+
+    private
+
+      # @return [Faraday::Connection]
+      def connection
+        @connection ||= Faraday.new(url: host) do |faraday|
+          faraday.use Faraday::Response::RaiseError
+          faraday.adapter :httpclient
+        end
+      end
+
+      # @return [ActionView::Base] used to render as though in an rails controller
+      def renderer
+        @renderer ||= begin
+          view_paths = ActionView::PathSet.new([File.dirname(__FILE__)])
+          ActionView::Base.new(view_paths, {})
+        end
+      end
+
+      # @param [Array<String>] fields
+      # @return [String]
+      def request_body(ids, fields)
+        renderer.render(
+          file: 'links_request.xml',
+          layout: false,
+          locals: {
+            client: self,
+            return_fields: fields,
+            cites: ids
+          }
+        )
+      end
+
+      # @param response [Faraday::Response]
+      def response_parse(response)
+        ng = Nokogiri::XML(response.body) { |config| config.strict.noblanks }.remove_namespaces!
+        pairs = ng.xpath('response/fn/map/map').map do |node|
+          [node.attr('name').split('_', 2).last, vals_to_hash(node.children.xpath('val'))]
```

Comment:
There is a lot going on here. Worth breaking parts of this into smaller private methods?